### PR TITLE
Uniswap V2 Integration

### PR DESCRIPTION
**Pull request by jinchung**

### CODE REVIEW

File: src/hooks/useUniswapPairs.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  const { allPairCombinations, calls } = useUniswapCalls(
+    inputCurrency,
+    outputCurrency
+  );
+
+  const { multicallResults } = useMulticall(
+    calls,
+    PAIR_INTERFACE,
+    PAIR_GET_RESERVES_FRAGMENT
+    // latestBlockNumber // TODO JIN
```
Comment (osdnk) [inline src/hooks/useUniswapPairs.ts:21]: how about this?

File: src/hooks/useUniswapCalls.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import { useMemo } from 'react';
+import { getTokenForCurrency } from '../handlers/uniswap';
+import {
+  PAIR_GET_RESERVES_CALL_DATA,
+  UNISWAP_V2_BASES,
+} from '../references/uniswap';
+
+import useAccountSettings from './useAccountSettings';
+
+export default function useUniswapCalls(inputCurrency, outputCurrency) {
```
Comment (brunobar79) [inline src/hooks/useUniswapCalls.ts:12]: typings

File: src/hooks/useUniswapCalls.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        outputToken,
+        base,
+      ]),
+      // each base against all bases
+      ...flatMap(bases, (base): [Token, Token][] =>
+        bases.map(otherBase => [base, otherBase])
+      ),
+    ];
+
+    let validCombos = filter(
```
Comment (brunobar79) [inline src/hooks/useUniswapCalls.ts:51]: const?

File: src/hooks/useUniswapPairs.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import { compact } from 'lodash';
+import { useMemo } from 'react';
+import {
+  PAIR_GET_RESERVES_FRAGMENT,
+  PAIR_INTERFACE,
+} from '../references/uniswap';
+import useMulticall from './useMulticall';
+import useUniswapCalls from './useUniswapCalls';
+
+export default function useUniswapPairs(inputCurrency, outputCurrency) {
```
Comment (brunobar79) [inline src/hooks/useUniswapPairs.ts:11]: typings

File: src/model/wallet.ts [hunk 5]
```diff
@@ -169,7 +166,10 @@ export const publicAccessControlOptions = {
 
 export function generateSeedPhrase(): EthereumMnemonic {
   logger.sentry('Generating a new seed phrase');
-  return ethers.utils.HDNode.entropyToMnemonic(ethers.utils.randomBytes(16));
+  console.log('HI - generate seed phrase');
```
Comment (brunobar79) [inline src/model/wallet.ts:169]: remove

File: src/model/wallet.ts [hunk 6]
```diff
@@ -169,7 +166,10 @@ export const publicAccessControlOptions = {
...
+  const seedPhrase = entropyToMnemonic(randomBytes(16));
+  console.log('HI - generate seed phrase', seedPhrase);
```
Comment (brunobar79) [inline src/model/wallet.ts:171]: remove

File: src/utils/contract.js [hunk 7]
```diff
@@ -22,7 +22,9 @@ const estimateApproveWithExchange = async (spender, exchange) => {
 };
 
 const estimateApprove = (tokenAddress, spender) => {
-  const exchange = new ethers.Contract(tokenAddress, erc20ABI, web3Provider);
+  console.log('HI - gonna make a contract', tokenAddress);
```
Comment (brunobar79) [inline src/utils/contract.js:25]: remove

File: src/utils/contract.js [hunk 8]
```diff
@@ -22,7 +22,9 @@ const estimateApproveWithExchange = async (spender, exchange) => {
...
+  const exchange = new Contract(tokenAddress, erc20ABI, web3Provider);
+  console.log('HI - exchnage contract', exchange);
```
Comment (brunobar79) [inline src/utils/contract.js:27]: remove

File: src/utils/contract.js [hunk 9]
```diff
@@ -35,15 +37,14 @@ const approve = async (
...
-  const exchange = new ethers.Contract(tokenAddress, erc20ABI, walletToUse);
-  const approval = await exchange.approve(
-    spender,
-    ethers.constants.MaxUint256,
-    {
-      gasLimit: gasLimit ? toHex(gasLimit) : undefined,
-      gasPrice: gasPrice ? toHex(gasPrice) : undefined,
-    }
-  );
+  console.log('HI - gonna make a contract', tokenAddress);
```
Comment (brunobar79) [inline src/utils/contract.js:40]: remove

File: src/utils/contract.js [hunk 10]
```diff
@@ -35,15 +37,14 @@ const approve = async (
...
+  const exchange = new Contract(tokenAddress, erc20ABI, walletToUse);
+  console.log('HI - exchange contrac tagain', exchange);
```
Comment (brunobar79) [inline src/utils/contract.js:42]: remove

File: src/utils/contract.js [hunk 11]
```diff
@@ -35,15 +37,14 @@ const approve = async (
...
+  const approval = await exchange.approve(spender, MaxUint256, {
+    gasLimit: gasLimit || undefined,
+    gasPrice: gasPrice || undefined,
+  });
+  console.log('HI - approval', approval);
```
Comment (brunobar79) [inline src/utils/contract.js:47]: remove

File: src/utils/contract.js [hunk 12]
```diff
@@ -52,11 +53,9 @@ const approve = async (
 
 const getRawAllowance = async (owner, token, spender) => {
   const { address: tokenAddress } = token;
-  const tokenContract = new ethers.Contract(
-    tokenAddress,
-    erc20ABI,
-    web3Provider
-  );
+  console.log('HI - creating token contract', tokenAddress);
```
Comment (brunobar79) [inline src/utils/contract.js:56]: remove

File: src/utils/contract.js [hunk 13]
```diff
@@ -52,11 +53,9 @@ const approve = async (
...
+  const tokenContract = new Contract(tokenAddress, erc20ABI, web3Provider);
+  console.log('HI - token contract', tokenContract);
```
Comment (brunobar79) [inline src/utils/contract.js:58]: remove

File: src/raps/unlockAndSwap.js [hunk 14]
```diff
@@ -15,60 +13,42 @@ import { createNewAction, createNewRap, RapActionTypes } from './common';
...
   if (!outputAmount) outputAmount = 1;
 
   const isValid = isValidSwapInput({
     inputCurrency,
-    inputReserve,
     outputCurrency,
-    outputReserve,
   });
 
   if (!isValid) return ethUnits.basic_swap;
```
Comment (brunobar79) [inline src/raps/unlockAndSwap.js:28]: should we increase this number? AFAIK some specific swaps are more expensive on v2

**Review by brunobar79 — COMMENTED** [review]
Looks great!  Most of the comments I added are to help the clean up process.
Will approve once e2e tests are passing, maybe @skylarbarrera  can help you with that.

**Comment by osdnk** [comment]
I'm sorry, I accidentally resolved conflict on this branch. Wanted to do it on another one. Hope it won't  break anything 🙏

**Comment by brunobar79** [comment]
FYI:  Build is failing because the Patch file: `patches/@ethersproject+providers+5.0.9.patch` is no longer valid, so you need to regenerate it or lock the version.

File: src/raps/unlockAndSwap.js (hunk 14 above)
Comment (jinchung) [reply src/raps/unlockAndSwap.js:28]: very good point!

**Comment by christianbaroni** [comment]
@jinchung tried sending ETH from send flow and got “failed to send transaction”

then installed latest release and worked fine

**Comment by christianbaroni** [comment]
@jinchung numbers in thousands missing commas

B5EF039A-D77F-4B4D-AE93-A59DFEE2A96A

**Comment by skylarbarrera** [comment]
3 issues I found regarding Swap:

Search Results are extremely slow: https://cloud.skylarbarrera.com/RPReplay_Final1603304097.mov 

Output input: 
- does not updated other inputs: https://cloud.skylarbarrera.com/RPReplay_Final1603304938.mov
- App crashes if updated while other inputs are filled: https://cloud.skylarbarrera.com/RPReplay_Final1603304376.mov

**Review by brunobar79 — APPROVED** [review]
LET'S GOOOOOOOO

**Review by skylarbarrera — APPROVED** [review]
🥳  🎉  Press the Button!

**Pull request merged**